pkg/odo/cli/init/init.go

```diff
@@ -0,0 +1,306 @@
+	if err != nil {
+		return err
+	}
+
+	if o.InitParams.Starter != "" {
+		// WARNING: this will remove all the content of the destination directory, ie the devfile.yaml file
+		err = o.downloadStarterProject(devfileObj, o.InitParams.Starter, o.destDir)
+		if err != nil {
+			return fmt.Errorf("unable to download starter project %q: %w", o.InitParams.Starter, err)
+		}
+		starterDownloaded = true
+
+		// in case the starter project contains a devfile, read it again
+		if _, err = o.fsys.Stat(destDevfile); err == nil {
+			devfileObj, err = devfile.ParseAndValidateFromFile(destDevfile)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// Set the name in the devfile *AND* writes the devfile back to the disk in case
+	// it has been removed and not replaced by the starter project
+	err = devfileObj.SetMetadataName(o.InitParams.Name)
+	if err != nil {
+		return fmt.Errorf("Failed to update the devfile's name: %w", err)
+	}
+
+	log.Italicf(`Your new component %q is ready in the current directory.
+To start editing your component, use "odo dev" and open this folder in your favorite IDE.
+Changes will be directly reflected on the cluster.
+To deploy your component to a cluster use “odo deploy”.`, o.InitParams.Name)
+
+	return nil
+}
+
+// downloadDirect downloads a devfile at the provided URL and saves it in dest
+func (o *InitOptions) downloadDirect(URL string, dest string) error {
+	parsedURL, err := url.Parse(URL)
+	if err != nil {
+		return err
+	}
+	if strings.HasPrefix(parsedURL.Scheme, "http") {
+		downloadSpinner := log.Spinnerf("Downloading devfile from %q", URL)
+		defer downloadSpinner.End(false)
+		params := util.HTTPRequestParams{
+			URL: URL,
+		}
+		devfileData, err := util.DownloadFileInMemory(params)
+		if err != nil {
+			return err
+		}
+		err = o.fsys.WriteFile(dest, devfileData, 0644)
+		if err != nil {
+			return err
+		}
+		downloadSpinner.End(true)
+	} else {
+		downloadSpinner := log.Spinnerf("Copying devfile from %q", URL)
+		defer downloadSpinner.End(false)
+		content, err := o.fsys.ReadFile(URL)
+		if err != nil {
+			return err
+		}
+		info, err := o.fsys.Stat(URL)
+		if err != nil {
+			return err
+		}
+		err = o.fsys.WriteFile(dest, content, info.Mode().Perm())
+		if err != nil {
+			return err
+		}
+		downloadSpinner.End(true)
+	}
+
+	return nil
+}
+
+// downloadRegistry downloads a devfile from the provided registry and saves it in dest
+func (o *InitOptions) downloadRegistry(registryName string, devfile string, dest string) error {
+	downloadSpinner := log.Spinnerf("Downloading devfile %q from registry %q", devfile, registryName)
+	defer downloadSpinner.End(false)
+	registries := o.preferenceClient.RegistryList()
+	var registry preference.Registry
+	var found bool
+	for _, registry = range *registries {
+		if registry.Name == registryName {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return fmt.Errorf("unable to find the registry with name %q", registryName)
+	}
```

Comment:
We are already validating in the `Validate` method if the registry exists. Aren't we repeating ourselves here? I'm pointing to the validation done by `o.InitParams.Validate(o.preferenceClient)`.

Reply:
But in any case, we have to protect from executing code if not found, or this piece of code out of context will look buggy. You can see this test as an assertion: "It should not fail, but let check it again, just in case of some bug."

Reply:
OK. Makes sense.